Sources/Rendering/OpenGL/Texture/index.js

```diff
@@ -1120,17 +1135,149 @@ function vtkOpenGLTexture(publicAPI, model) {
 
     const haveWebgl2 = model.openGLRenderWindow.getWebgl2();
 
+    let reformatGradientsFunction;
+    if (haveWebgl2) {
+      reformatGradientsFunction = (workerResults) => {
+        const numVoxelsIn = width * height * depth;
+        const reformattedGradients = new Uint8Array(numVoxelsIn * 4);
+        const maxMag = model.volumeInfo.max;
+
+        workerResults.forEach(
+          ({
+            subGradients,
+            subMagnitudes,
+            subMinMag,
+            subMaxMag,
+            subDepthStart,
+            subDepthEnd,
+          }) => {
+            let inIdx = 0;
+            let inMagIdx = 0;
+            let outIdx = subDepthStart * width * height * 4;
+            // start and end depths are inclusive
+            const numWorkerVoxels =
+              width * height * (subDepthEnd - subDepthStart + 1);
+            for (let vp = 0; vp < numWorkerVoxels; ++vp) {
+              reformattedGradients[outIdx++] = subGradients[inIdx++];
+              reformattedGradients[outIdx++] = subGradients[inIdx++];
+              reformattedGradients[outIdx++] = subGradients[inIdx++];
+              reformattedGradients[outIdx++] =
+                255.0 * Math.sqrt(subMagnitudes[inMagIdx++] / maxMag);
+            }
+          }
+        );
+
+        return publicAPI.create3DFromRaw(
+          width,
+          height,
+          depth,
+          4,
+          VtkDataTypes.UNSIGNED_CHAR,
+          reformattedGradients
+        );
+      };
+    } else {
+      // Now determine the texture parameters using the arguments.
+      publicAPI.getOpenGLDataType(VtkDataTypes.UNSIGNED_CHAR);
+      publicAPI.getInternalFormat(VtkDataTypes.UNSIGNED_CHAR, 4);
+      publicAPI.getFormat(VtkDataTypes.UNSIGNED_CHAR, 4);
+
+      if (!model.internalFormat || !model.format || !model.openGLDataType) {
+        vtkErrorMacro('Failed to determine texture parameters.');
+        return;
```

Comment:
What are the general design principles about exception handling in vtk.js? Would it be better to throw an exception instead of just logging and returning?

Reply:
I don't really know, this sure looks very C++y. @jourdain @martinken ?

Reply:
Use the macro for now. We can make those macro throw an exception if we want to go that path.

Reply:
Referenced in this issue: https://github.com/Kitware/vtk-js/issues/640